catalyst-gateway-crates/c509-certificate/src/c509_cert.rs

```diff
@@ -0,0 +1,282 @@
+// ---------------------------------------------------
+
+/// Encodes the subject public key information.
+///
+/// # Arguments
+///
+/// * `pubk_type` - The type of the public key algorithm.
+/// * `pubk` - The public key bytes.
+/// * `encoder` - A mutable reference to a `minicbor::Encoder` to encode the extensions
+///   into.
+///
+/// # Panics
+///
+/// Panics if the public key type is not supported (RSA and Elliptic curve in Weierstraß
+/// public keys).
+///
+/// # References
+///
+/// Please refer to the [c509-certificate](https://datatracker.ietf.org/doc/draft-ietf-cose-cbor-encoded-cert/09/) Section 3.1. Message Fields -> subjectPublicKeyInfo for more information.
+fn encode_subject_public_key(
+    pubk_type: PubKeyAlgoRegistry, pubk: Vec<u8>, encoder: &mut Encoder<&mut Vec<u8>>,
+) {
+    // RSA public keys
+    if pubk_type == PubKeyAlgoRegistry::RSA {
+        panic!("Not support yet");
+    // Elliptic curve public keys in Weierstraß
+    } else if [
+        PubKeyAlgoRegistry::Secp256r1,
+        PubKeyAlgoRegistry::Secp384r1,
+        PubKeyAlgoRegistry::Secp521r1,
+        PubKeyAlgoRegistry::Brainpool256r1,
+        PubKeyAlgoRegistry::Brainpool384r1,
+        PubKeyAlgoRegistry::Brainpool512r1,
+        PubKeyAlgoRegistry::Frp256v1,
+        PubKeyAlgoRegistry::SM2P256v1,
+    ]
+    .contains(&pubk_type)
+    {
+        panic!("Not support yet");
+    } else {
+        let _unused = encoder.i8(pubk_type as i8);
+        let _unused = encoder.bytes(&pubk);
+    }
+}
+
+// ---------------------------------------------------
+
+/// Encodes the signature algorithm information.
+///
+/// # Arguments
+///
+/// * `sig_algo` - The signature algorithm.
+/// * `encoder` - A mutable reference to a `minicbor::Encoder` to encode the extensions
+///   into.
+/// Please refer to the [c509-certificate](https://datatracker.ietf.org/doc/draft-ietf-cose-cbor-encoded-cert/09/) Section 3.1. Message Fields -> signatureAlgorithm for more information.
+fn encode_signature_algorithm(
+    sig_algo: SignatureAlgoRegistry, encoder: &mut Encoder<&mut Vec<u8>>,
+) {
+    let _unused = encoder.i8(sig_algo as i8);
+}
+
+// ---------------------------------------------------
+
+/// Encodes the issuer signature value.
+///
+/// # Arguments
+///
+/// * `_sig_value` - The signature value (currently unused).
+/// * `encoder` - A mutable reference to a `minicbor::Encoder` to encode the extensions
+///   into.
+///
+/// # Remarks
+/// This function will be implemented in the future.
+/// Please refer to the [c509-certificate](https://datatracker.ietf.org/doc/draft-ietf-cose-cbor-encoded-cert/09/) Section 3.1. Message Fields -> signatureValue for more information.
+fn encode_issuer_signature_value(_sig_value: Vec<u8>, encoder: &mut Encoder<&mut Vec<u8>>) {
+    // Currently will generate an unsigned C509 certificates,
+    // hence the value is null
+    let _unused = encoder.null();
+}
+
+// ---------------------------------------------------
+
+#[wasm_bindgen]
```

Comment:
I think all macro definitions better to place after comments, so we will not split the Rust code logic.
```suggestion
```